### PR TITLE
Suite of new conversion and maths based functions

### PR DESCRIPTION
**Pull request by breezy-codes**
# Description

This pull request contains a suite of conversion functions for converting between different values, including input validation for these. In addition to 4 maths based functions.

The added functions are - 
- For input validation - `is_binary`, `is_hex` and `is_octal`
- For conversions -  `dec_to_bin` and back, `hex_to_bin` and back, `oct_to_dec` and back, `hex_to_ct` and back, `base64` encoding and decoding.
- For maths - `calculate_square_root`, `is_prime_number`, `greatest_common_divisor` and `least_common_multiple`.

The motivation for these new functions was based on the addition of some maths and cryptography based tutorials, these are great for converting values for encryption as well as useful in the upcoming steganography tutorials, the maths ones build foundation points for numerous other maths operations and will benefit the maths/cryptography tutorials too.

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

I've done extensive unit checks for checking for both correct and incorrect values as well as edge cases. The unit tests are included in this PR. Some of the unit tests also output when running `sktest` as further validation.

## Testing Checklist

- [X] Tested with sktest
- [X] Tested with skunit_tests

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings

image

### CODE REVIEW

**Review by matt439 — CHANGES_REQUESTED** [review]
Everything works very well. I tested the base 64 encoding/decoding with special characters and it handled them flawlessly. The string validity functions also work properly and I was unable to find any edge cases which were not detected. `calculate_square_root` correctly returns `nan` when passing a negative value.

Some minor description changes for some of the functions will help students debug their programs if they are supplying the functions with invalid strings.  I would consider printing warnings to the terminal when invalid strings are used too, so that its even more obvious that they are making a mistake. I tested everything on WSL.

File: coresdk/src/coresdk/basics.cpp [hunk 1]
```diff
@@ -14,10 +14,14 @@
 #include <functional>
 #include <cctype>
 #include <locale>
+using std::to_string;
```
[the remark this answers is not in the document]
Comment (breezy-codes) [reply coresdk/src/coresdk/basics.cpp:17]: It is used in some other files but I'll remove and swap to manually doing it :)

File: coresdk/src/coresdk/basics.h [hunk 2]
```diff
@@ -135,6 +135,230 @@ namespace splashkit_lib
...
+
+    /**
+     * @brief Converts a binary string to a decimal (unsigned integer)
+     *
+     * Converts the provided binary string into an unsigned integer.
+     * For example, "1010" will be converted to 10.
+     *
+     * @param bin Binary string to convert
+     *
+     * @returns Decimal (unsigned integer) representation of the binary string
```
[the remark this answers is not in the document]
Comment (breezy-codes) [reply coresdk/src/coresdk/basics.h:193]: Made this change

**Review by matt439 — APPROVED** [review]
The new changes look good and have resolved what I raised in comments.

File: coresdk/src/test/test_terminal.cpp [hunk 3]
```diff
@@ -27,7 +29,866 @@ void test_simple_terminal()
...
+    assert(dec_to_bin(123) != "1111100");
+    assert(dec_to_bin(1025) != "10000000010");
+    assert(dec_to_bin(255) != "100000000");
+    assert(dec_to_bin(256) != "100000001");
+    assert(dec_to_bin(1023) != "10000000000");
+    assert(dec_to_bin(4096) != "1000000000001");
+    assert(dec_to_bin(4294967295) != "1000");
+    assert(dec_to_bin(2147483648) != "1000001");
+    assert(dec_to_bin(123456789) != "1110101");
+    assert(dec_to_bin(1100) != "11110");
```
Comment (WhyPenguins) [inline coresdk/src/test/test_terminal.cpp:76]: Same as the comment on left on #79, I don't think these negative tests don't really add anything - if the functions are bugged, the likelihood they result in these specific wrong outputs is very low; additionally, many of these were already `assert`ed to be working correctly further up, so if we've reached this point in the code then they _can't_ fail.
This applies to all the other test functions (won't duplicate this message onto them though :rofl:)

File: coresdk/src/test/test_terminal.cpp [hunk 4]
```diff
@@ -27,7 +29,866 @@ void test_simple_terminal()
...
+    assert(hex_to_bin("F") != "10000");
+    assert(hex_to_bin("10") != "00010001");
+    assert(hex_to_bin("FF") != "100000000");
+    assert(hex_to_bin("100") != "000100000001");
+    assert(hex_to_bin("ABC") != "101010111101");
+    assert(hex_to_bin("1234") != "0001001000110101");
+    assert(hex_to_bin("FFFF") != "11111111111111110");
+    assert(hex_to_bin("FFFFFFFF") != "11111");
+    assert(hex_to_bin("1FFFFFF") != "0111110");
+    assert(hex_to_bin("GGGGGG") != "10101011110010");
```
Comment (WhyPenguins) [inline coresdk/src/test/test_terminal.cpp:182]: This applies to all the others like this as well, but if it's testing invalid input that should be handled a specific way (in this case returning an empty string), then _that's_ what should be tested for. There are _millions_ of possible things we don't want it to output, but only one thing we do. `assert(hex_to_bin("GGGGGG") == "");` would be much better.

**Review by WhyPenguins — COMMENTED** [review]
This looks really comprehensive, nice work! The math functions and base64 conversion parts look great :smiley:  

I do think the conversion functions could be a lot simpler though, which makes me unsure whether to mark as "Approve". I'll leave it as 'Comment' rather than 'Request Changes', since it is a substantial addition and it might be good to include anyway haha.

Main thing for the conversion part is that I feel there's _too_ much work done haha. There were much easier ways to do a lot of this - for instance all of the `**_to_dec` functions could have been basically one-liners that use `stoi` with the base overload like you did in #79; similarly most of the `dec_to_**` could have been done with stringstream (also like in #79). 

Then the other `a_to_b` could all have been one-liners like `return dec_to_b(a_to_dec(str));`, like you have in `hex_to_oct` and `oct_to_hex` :smiley: 

This would substantially reduce the amount of code (and therefore number of potential bugs), so while it's a shame considering how much work was done, I feel like simplifying those functions would be better for the code-base in the long run. It'd also allow negative numbers as a side-effect haha, which right now aren't supported at all.

The tests also really should be in the unit tests (these are the sorts of tests that are perfect for it! :smiley:). I'm not sure why they're in `test_terminal.cpp` to be honest, a new file would have been better.

Overall good work though! In particular I like the base64 implementations, those are pretty nice haha.

**Comment by macite** [comment]
Fixed (some) and merged
